zmon_aws_agent/main.py

```diff
@@ -0,0 +1,181 @@
+        try:
+            response = requests.get('http://169.254.169.254/latest/meta-data/placement/availability-zone', timeout=2)
+        except:
+            logger.error('Region was not specified as a parameter and can not be fetched from instance meta-data!')
+            raise
+        region = response.text[:-1]
+    else:
+        region = args.region
+
+    logger.info('Using region: {}'.format(region))
+
+    logger.info('Entity service URL: %s', args.entityservice)
+
+    logger.info('Reading DNS data for hosted zones')
+    aws.populate_dns_data()
+
+    # 2. Get running apps
+    apps = aws.get_running_apps(region)
+
+    infrastructure_account = None
+
+    elbs = []
+    scaling_groups = []
+    rds = []
+    elasticaches = []
+    dynamodbs = []
+
+    new_entities = []
+    to_be_removed = []
+
+    if len(apps) > 0:
+        infrastructure_account = apps[0]['infrastructure_account']
+        elbs = aws.get_running_elbs(region, infrastructure_account)
+        scaling_groups = aws.get_auto_scaling_groups(region, infrastructure_account)
+        rds = aws.get_rds_instances(region, infrastructure_account)
+        elasticaches = aws.get_elasticache_nodes(region, infrastructure_account)
+        dynamodbs = aws.get_dynamodb_tables(region, infrastructure_account)
+
+    if infrastructure_account:
+        account_alias = aws.get_account_alias(region)
+        ia_entity = {
+            'type': 'local',
+            'infrastructure_account': infrastructure_account,
+            'account_alias': account_alias,
+            'region': region,
+            'id': 'aws-ac[{}:{}]'.format(infrastructure_account, region),
+            'created_by': 'agent',
+        }
+
+        application_entities = aws.get_apps_from_entities(apps, infrastructure_account, region)
+
+        current_entities = elbs + scaling_groups + apps + application_entities + rds + elasticaches + dynamodbs
+
+        # 3. ZMON entities
+        token = None if args.disable_oauth2 else os.getenv('ZMON_AGENT_TOKEN', tokens.get('uid'))
+        zmon_client = ZMon(args.entityservice, token=token)
+
+        query = {'infrastructure_account': infrastructure_account, 'region': region, 'created_by': 'agent'}
+        entities = zmon_client.get_entities(query)
+
+        # 4. Removing misssing entities
+        existing_ids = get_existing_ids(entities)
+        current_entities_ids = [e['id'] for e in current_entities]
+
+        to_be_removed = remove_missing_entities(existing_ids, current_entities_ids, zmon_client, json=args.json)
+
+        # 5. Always add Local entity
+        if not args.json:
+            zmon_client.add_entity(ia_entity)
+
+        # 6. Get new/updated entities
+        new_entities = add_new_entities(current_entities, entities, zmon_client, json=args.json)
+
+        logger.info('Found {} new entities from {} entities'.format(len(new_entities), len(current_entities)))
+
+    # Check if it is a dry run!
+    if args.json:
+        d = {
```

Comment:
I just want to prevent repeating the entity types everywhere --- adding a new entity type (e.g. for a new AWS product) should be as easy as adding one "collector" method somewhere which returns the new entities